src/service/mod.rs

```diff
@@ -59,3 +59,35 @@ impl Reconnect {
         self.next_time = Instant::now() + backoff;
     }
 }
+
+pub struct AckTimer {
+    next_time: Instant,
+    timeout: Duration,
+}
+
+impl AckTimer {
+    pub fn new(timeout: Duration) -> Self {
+        Self {
+            next_time: Instant::now() + timeout,
+            timeout,
+        }
+    }
+
+    pub async fn wait(&self) {
+        if self.next_time >= Instant::now() {
+            time::sleep_until(self.next_time).await
+        } else {
+            std::future::pending().await
```

Comment:
This is surprising to see. Is it necessary? The reconnect timer has only the first branch.

Reply:
Oh, is it to briefly give the scheduler a chance to do something else?

Reply:
> This is surprising to see. Is it necessary? The reconnect timer has only the first branch.

It's there in case the ack_timeout is not needed at all (i.e. 0).. having a next_time in the past will just cause the sleep to fire immediately. Using a pending will just have it be a no-op await in the packet router Tokio select